### ISSUE TITLE
Social sharing of pub packages not as rich as sharing a github repo

### ISSUE THREAD

**Issue by sethladd** [dart-lang/pub-dev#7886]
We noticed that there are some opportunities to improve the social sharing experience of a pub package hosted on pub.dev. Specifically: more auto-generated context and "splash image" of the pub package.

In comparison, here is what it looks like to share a github project (via X):

![Screenshot 2024-07-22 at 10 23 20 AM](https://github.com/user-attachments/assets/e24ca702-16ea-4210-bef5-afb042f31a24)

Note the large image with content around what the package is, contributors, stars, forks, etc (giving a more accurate impression about the status of the project).

Compare that with a social share from the same project's pub package:

![Screenshot 2024-07-22 at 10 23 01 AM](https://github.com/user-attachments/assets/51f9e3eb-44d4-45e8-889f-f53aa6bb632b)

Note the static stock image for "pub", with no package branding/identification/stats

To help people promote their awesome work as hosted on pub.dev, we should consider how to make a more rich "social share snippet image". We can use GitHub as the inspiration here.

Ideally, when sharing a package's pub.dev URL via social services like X, it would be good to:

* auto-generate a snippet image
* including the name of the package
* including any primary image from the package
* including a one-line description of the package
* including some basic stats like stars, pub score

Thanks for considering!

**Comment by sigurdm** [comment]
Seems like for twitter we need some meta tags: https://developer.x.com/en/docs/twitter-for-websites/cards/guides/getting-started

Google chat seems to use opengraph (hinted here https://webapps.stackexchange.com/questions/151458/is-it-possible-to-change-the-image-google-chat-shows-for-a-url-link): https://ogp.me/ (`og:image`, `og:description`)

**Comment by sigurdm** [comment]
Seems we are already creating these tags, but the content is not as precise as wanted.

https://github.com/dart-lang/pub-dev/blob/bf3d7029ebcdb5ec9ab7bb045f19ed0d6124f792/app/lib/frontend/templates/views/shared/layout.dart#L75

We could specialize the image if there is one in the readme or perhaps use one of the screenshots.

https://robolly.com/open-graph-preview/?url=https://pub.dev/packages/retry

![image](https://github.com/user-attachments/assets/d111470d-17b4-4a59-8bb8-5d193b7e132b)

**Comment by halildurmus** [comment]
I believe the best approach would be adding a new field to the `pubspec.yaml`, for example `social_image`, which would allow package authors to specify a preferred image for social previews. If omitted, an image from the screenshots or one embedded in the README could be used. In cases where no images are available, pub.dev could generate a preview image, similar to how GitHub does.

**Comment by isoos** [comment]
I think we shouldn't add yet another tag into `pubspec.yaml` for this, as if you have a screenshot, you already order them by importance, and the first one will be the one displayed on the package listing pages. We should just reuse it for social media sharing.

**Comment by sigurdm** [comment]
> and the first one will be the one displayed on the package listing pages.

However, the intended semantics of screenshots is to be ... screenshots - not clear that that is the same as a display image.

**Comment by sigurdm** [comment]
But I agree, adding yet another pubspec property just for this also sucks

**Comment by halildurmus** [comment]
> > and the first one will be the one displayed on the package listing pages.
> 
> However, the intended semantics of screenshots is to be ... screenshots - not clear that that is the same as a display image.

+1 to this. Also, the [documentation](https://dart.dev/tools/pub/pubspec#screenshots) of the `screenshots` property states:
> Don't include logos or other branding imagery in this section.

**Comment by FMorschel** [comment]
I prefer a specific and _explicit_ place for us to place that if not on `pubspec.yaml` directly, maybe somewhere on the actual website (it feels weird since nothing else is done through it).

**Comment by dickermoshe** [comment]
Instead of users directly controlling the entire image, why not use a style similar to Githubs social image which shows a nice graphic with test as shown in the issue above.

I would work on this myself, my only question is:

1. What package would be acceptable for creating images. There is a maintained package of [opencv](https://pub.dev/packages/dartcv4) on pub. Although I doubt you would feel comfortable using it.
2. A separate API would need to be created to generating these image. `/genimage?package=drift` which would generate a new image if the package stats changed from the last request.
3. How would we go about caching this image
4. What information should we show? (duh)

**Comment by sethladd** [comment]
Yes, I was assuming that if this was done it would be similar to how Github does it:

<img width="632" alt="Image" src="https://github.com/user-attachments/assets/7e0218fc-b28e-497f-9b60-22e3ca18f680" />

For (4), see the original issue description above: https://github.com/dart-lang/pub-dev/issues/7886#issue-2423390180

The default social share image would be some auto-generated image of the package name, purpose, and other stats from pub. This can always be overridden if the package wants to provide a specified image to use instead.

**Comment by sigurdm** [comment]
Just for reference: in github this seems to be configured outside the repo: https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/customizing-your-repositorys-social-media-preview

**Comment by sigurdm** [comment]
The screenshots are uploaded from the worker which is running pana. See perhaps: https://github.com/dart-lang/pana/blob/master/lib/src/screenshots.dart

**Comment by isoos** [comment]
> Does this repo contain the code which uploads screenshots?

Screenshots are extracted via `pana` (see link above) and uploaded as part of the analysis blob. I think the screenshots/backend method you've linked to is now obsolete. (a question to @szakarias)

The related currently used code is in `task/backend.dart` and also in `routes.dart` under the `.../gen-res/...` handler.

**Comment by isoos** [comment]
> I will DIY some opencv once I see a clear path for the infra

Please note that we would want to discuss the design (UI and backend) first, before accepting any PR. Providing good arguments and design is more helpful at this stage than any code.

**Comment by dickermoshe** [comment]
## Splitting the Issue

I believe this issue can be divided into two distinct features:

1. Providing an appealing infographic when pub.dev links are shared on social media
2. Allowing package authors to customize these graphics

### Reasons for splitting:

Based on my experience, very few package authors customize social media previews even when given the option. Looking at the top 10 starred repositories, none customize this feature. Therefore, I don't think customization should be a prerequisite for implementing the basic functionality.

Additionally, the customization component significantly increases complexity compared to the original issue:
- Modifying the pubspec format would create friction for minimal benefit
- Adding a UI to the website introduces complexity to what is currently a streamlined interface
- Most package customizations are already handled through the pubspec format; adding another interface specifically for social previews seems inconsistent with the platform's design philosophy

Since this feature would be used by relatively few authors but adds considerable complexity, I recommend implementing the basic infographic first, then evaluating whether customization options are truly needed based on user feedback.

**Comment by sethladd** [comment]
> Just for reference: in github this seems to be configured outside the repo: https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/customizing-your-repositorys-social-media-preview

thanks! Yeah, that seems to be docs for how to replace the default social share card.

https://github.com/dart-lang/pub-dev/issues/7886#issuecomment-2833652378 has a screenshot of a default social share card from github, which seems to be auto generated to include stats like contributors, stars, forks, etc.

My wish or dream is that pub's default social share card is auto generated and contains info such as:

including the name of the package
including any primary image from the package
including a one-line description of the package
including some basic stats like stars, pub score
and including a one-liner for pub (like the bottom of the github social share card is a one-liner for github)

**Comment by sethladd** [comment]
Here's an example of what I'm thinking, for the default social share card coming out of pub. Just an example:

![Image](https://github.com/user-attachments/assets/f99a4baa-7bcb-4af8-8b8e-7ea6c374b5a1)

**Comment by dickermoshe** [comment]
@isoos 
I'm working on a design for how the backend would handle image generation. 

It seems now that `pub_worker` only handles a single payload which instructs it to rebuild a package.
This is a relatively expensive process and generating new images there would be unreasonable.

Generating new images will need to be done relatively frequently. Exactly how and when we decide that an image is stale would involve a couple of factors, but it would still be much more frequent than pub_worker can handle.

I will investigate how intensive `opencv` is, and if we can do it on the frontend server itself.
Otherwise we would either need to create another background worker, or adapt pub_worker to handle different types of payloads

1. Is there a more general purpose background-task api that pub.dev already has?
2. Can shelf do work after a response is returned? Maybe in an isolate?

**Comment by dickermoshe** [comment]
@sethladd needs better contrast on the bottom

**Comment by sethladd** [comment]
> [@sethladd](https://github.com/sethladd) needs better contrast on the bottom

ack, good catch. once we figure out how to do this, we can tweak the design.
